### PR TITLE
Feature/Block TCP connections with RST packet

### PR DESCRIPTION
**Pull request by vlabo**

### CODE REVIEW

File: sys/pm_callouts.c [hunk 1]
```diff
@@ -408,6 +408,191 @@ void redir(portmaster_packet_info* packetInfo, portmaster_packet_info* redirInfo
...
+        // Create the net buffer list for the new RST packet
+        status = wrap_packet_data_in_nb(tcpResetPacket, packetLength, &injectNBL);
+        if (!NT_SUCCESS(status)) {
+            ERR("send_tcp_rst ipv6 -> wrap_packet_data_in_nb failed: %u", status);
+            portmaster_free(tcpResetPacket);
+            return;
+        }
+
+        // loopback needs special treatment
+        if(is_ipv6_loopback(packetInfo->remoteIP)) {
```
Comment (dhaavi) [inline sys/pm_callouts.c:465]: This *should* only be happening for outgoing packets, but we may want to block an incoming packet.
I suggest to check for loopback, or incoming packets.

File: sys/pm_callouts.c [hunk 2]
```diff
@@ -408,6 +408,191 @@ void redir(portmaster_packet_info* packetInfo, portmaster_packet_info* redirInfo
...
+        // Create the net buffer list for the new RST packet
+        status = wrap_packet_data_in_nb(tcpResetPacket, packetLength, &injectNBL);
+        if (!NT_SUCCESS(status)) {
+            ERR("send_tcp_rst ipv4 -> wrap_packet_data_in_nb failed: %u", status);
+            portmaster_free(tcpResetPacket);
+            return;
+        }
+
+        // loopback needs special treatment
+        if(is_ipv4_loopback(packetInfo->remoteIP[0])) {
```
Comment (dhaavi) [inline sys/pm_callouts.c:535]: This _should_ only be happening for outgoing packets, but we may want to block an incoming packet.
I suggest to check for loopback, or incoming packets.

File: sys/pm_callouts.c [hunk 3]
```diff
@@ -408,6 +408,191 @@ void redir(portmaster_packet_info* packetInfo, portmaster_packet_info* redirInfo
...
+
+        // allocate memory for the reset packet
+        tcpResetPacket = portmaster_malloc(packetLength, FALSE);
+
+        // initialize IPv6 header
+        ipHeader = (PIPV6_HEADER) tcpResetPacket;
+        ipHeader->Version = 6;
+        ipHeader->Length = sizeof(TCP_HEADER);
+        ipHeader->NextHdr = 6;
+        ipHeader->HopLimit = 128;
```
Comment (dhaavi) [inline sys/pm_callouts.c:439]: Isn't Hop Limit an 8 bit field?

File: sys/pm_callouts.c [hunk 4]
```diff
@@ -408,6 +408,191 @@ void redir(portmaster_packet_info* packetInfo, portmaster_packet_info* redirInfo
...
+
+        // initialize IPv4 header
+        ipHeader = (PIPV4_HEADER) tcpResetPacket;        
+        ipHeader->HdrLength = sizeof(IPV4_HEADER) / 4;
+        ipHeader->Version = 4;
+        ipHeader->TOS = 0;
+        ipHeader->Length = RtlUshortByteSwap(packetLength);
+        ipHeader->Id = 0;
+        ipHeader->Protocol = packetInfo->protocol;
+        ipHeader->TTL = 128;
```
Comment (dhaavi) [inline sys/pm_callouts.c:509]: Isn't TTL an 8 bit field?

File: sys/pm_callouts.c [hunk 5]
```diff
@@ -408,6 +408,191 @@ void redir(portmaster_packet_info* packetInfo, portmaster_packet_info* redirInfo
...
+        PTCP_HEADER tcpHeader;
+
+        // allocate memory for the reset packet
+        tcpResetPacket = portmaster_malloc(packetLength, FALSE);
+
+        // initialize IPv6 header
+        ipHeader = (PIPV6_HEADER) tcpResetPacket;
+        ipHeader->Version = 6;
+        ipHeader->Length = sizeof(TCP_HEADER);
+        ipHeader->NextHdr = 6;
```
Comment (dhaavi) [inline sys/pm_callouts.c:438]: Use `packetInfo->protocol` insteads for more clarity. (Or, Add a comment to which protocol you are setting it. Yes it's obvious, but it's also not.)

**Review by dhaavi — CHANGES_REQUESTED** [review]
Very nice! Just some minor things.

File: sys/pm_callouts.c (hunk 3 above)
Comment (vlabo) [reply sys/pm_callouts.c:439]: Yes but it is unsigned int so max value is 255

Comment (dhaavi) [reply sys/pm_callouts.c:439]: Ah, yes. My bad.

**Review by dhaavi — APPROVED** [review]

**Pull request merged**